src/shared/type-coercions.js

```diff
@@ -67,3 +67,19 @@ export function toString(value) {
   }
   return '';
 }
+
+/**
+ * Coerce incoming value to appropriate `showHighlights` string values.
+ *   - true => 'always' OR
+ *   - false => 'never' OR
+ *   - the passed in value as a string
+ *
+ * @param {any} value - initial value
+ */
+export function toShowHighlights(value) {
+  // Convert legacy keys/values to corresponding current configuration.
+  if (typeof value === 'boolean') {
+    return value ? 'always' : 'never';
+  }
+  return toString(value);
```

Comment:
This is consistent with the previous behavior of the `showHighlights` getter, but not ideal IMO as the type of the `showHighlights` config setting is logically `'always'|'never'|'whenSidebarOpen'`. From an end user's point of view, I think the most useful behavior would be to:

1. Convert booleans to 'always'/'never'
2. Return valid enum values ('always', 'never', 'whenSidebarOpen') unchanged
3. Return 'always' if the config value is not specified (I think this is handled elsewhere via `defaultValue`)?
4. Return 'always' for any other value, but log a warning to inform the user that the value is not valid

One possible approach might be to implement this in a generic way by having the coerce methods throw a specific for invalid values (`InvalidValueError`?) which are then caught in higher level code, causing a fallback to the default value.

If you agree, then this could perhaps be split into two phases: 1) In this PR, just make `toShowHighlights` silently fall back to `always` if the value is not valid. 2) In a follow-up, introduce a mechanism to handle coercion errors more generically.

Reply:
I think we're trending towards validation which admittedly is needed.  I have an older PR with proposed validation methods but we never got around to evaluating it. It was which was originality for the sidebar, but could work for both here and the sidebar eventually.

https://github.com/hypothesis/client/commit/7537d3aba56ab9f297074631b9141eeb8e61aa4d

(take a look at the tests)

It's a bit too robust just a simple string, but it could be used to implement a thorough validation schema for our config. If I recall correctly, it can do multiple types with various conditions, even  objects and arrays. So in this case case, we could pass a boolean or a limited set of strings, and report a console error otherwise.

I think we may need to take a step back and think how we might include validation on all the config values because I think there is a lot of upside to reporting config problems to users as you just pointed out above. A simple console warn or error might be sufficient and perhaps if a config value fails validation, we could remove its value safely from the config

The big question is at what point to we validate vs coerce? I suspect doing the validation first might be the best way as we can report directly on the value that the user provides.

For now, I can do exactly what you suggested, we'll make the coerge silently mutate to "always", but then we should follow up with a plan for full config validation.  I think there especially may be value in report errors when config values need to be structured objects.

Reply:
> The big question is at what point to we validate vs coerce? I suspect doing the validation first might be the best way as we can report directly on the value that the user provides.

My general thought is that being too liberal with coercing values can be problematic because it is hard to know how much of that liberalness users are relying on and what might break if it becomes stricter in future. For that reason I favor being quite strict ie. doing mostly validation with implicit coercion limited to very specific cases, if any.

> For now, I can do exactly what you suggested, we'll make the coerge silently mutate to "always", but then we should follow up with a plan for full config validation. I think there especially may be value in report errors when config values need to be structured objects.

In this case any invalid values will have the same effect as not supplying the value at all, since "always" is the default value. This is what we would do if we only had validation and a policy of ignoring invalid values. This sounds good to me, as does opening an issue to implement more complete validation in future.

Reply:
gotcha, I follow